### ISSUE TITLE
Django

### ISSUE THREAD

**Issue by amille44420** [miguelgrinberg/python-socketio#25]
Hi ~ 

This is not an issue or anything, just an idea I'm interested in and I wanna talk about it. I work with Django but I often miss using WebSocket like Socket.IO does. So this module, python-socket.io is pretty interesting for me (and I should say this is an amazing job).

Anyway, I'm trying to find the best way to use this module with Django. First I was thinking about writing a Middleware then later I thought that it might be a better idea to let the developer handle the entry point if we could easily define that entry like a view using the url router provided by Django.

I tried something like that and it seems to work, I might have a wrong idea about how to mix the two together.

```python
@csrf_exempt
def socketio_view(request):
    # instance a new response
    response = HttpResponse(content_type='application/json')

    def start_response(status, headers):
        # inform the headers & status
        # first get the status code as integer
        response.status_code = int(status.split(' ')[0])
        # then inform headers
        for key, value in headers:
            response[key] = value

    # call the socket io request handler
    response.content = sio.handle_request(request.META, start_response)

    return response
```

I was thinking about setup the socket io server inside the app configuration (AppConfig) of the module.

**Comment by miguelgrinberg** [comment]
I include a middleware to dispatch HTTP traffic appropriately between the WSGI app and the SocketIO server. See how I use it in the Flask based example in this repo. You could handle the SocketIO traffic directly inside your application too, but I'm not sure there is a benefit to doing it that way.

**Comment by amille44420** [comment]
I didn't work previously with evenlet and I was wondering if I can really use eventlet knowing I run my Django apps with uWSGI (emperor mode, forked & threaded vassals) on production.

**Comment by miguelgrinberg** [comment]
@amille44420 No, uWSGI is not very friendly to SocketIO at this time. For your situation, my recommendation is that you start a standalone SocketIO server based on eventlet (or more than one if your needs demand it), and then communicate from your Django app via message queue.

**Comment by amille44420** [comment]
I had a look for using gevent with uwsgi but I couldn't verify it because I'm working with Py3. Right now gevent-websocket doesn't support Py3. However I was thinking about using gunicorn instead of uwsgi. I'll look into it later but by any chance do you know if gunicorn works with eventlet ? My first solution works and we can also set it up as you suggest (using the Middleware) but the transport won't upgrade to websocket in threading mode. That's why I went looking into gevent.

**Comment by miguelgrinberg** [comment]
@amille44420 I spent a considerable amount of time looking at all the different options. You can make uWSGI work with Flask-SocketIO using gevent, but you have to give up WebSocket for that. The problem is that gevent-websocket is not compatible with uWSGI, which comes with its own WebSocket implementation.

Gunicorn works really well with eventlet, but a requirement of Flask-SocketIO is that you using a single worker, because Gunicorn at this time does not support sticky sessions. You can start multiple Gunicorn processes, each with a single worker, and that will work fine (if you also start a message queue for the SocketIO servers to communicate).

**Comment by amille44420** [comment]
Thank you very much for your help and your works.

I did use python-socketio in my Django project.

- I use eventlet with python-socketio
- WSGI_APPLICATION is a Middleware instance (the one provided by python-socketio)
- I use the monkey patch of eventlet in both the django manager (manage.py) and when I initialize my wsgi instance (but I only do the patch for the runserver command for my manager). It doesn't matter to patch twice (like the documentation of eventlet says)
- I override the command runserver to replace the run method (django.core.servers.basehttp.run) by one starting a wsgi server with eventlet instead (I do it a little bit like the django.contrib.staticfiles app does). However I still have to patch in manage.py to support eventlet.
- I'm working with PyCharm so I had to turn on the gevent support to make everything works with the debugger

It's very important to use the monkey patch as early as possible. Otherwise no errors will raise but the messages send with python-socketio in threads won't reach the client. (we must do like you did here https://github.com/miguelgrinberg/python-socketio/blob/master/examples/app.py).

Anyhow, maybe what I did is not perfect yet but it works well. However I add to use PgBouncer (transaction pooling) otherwise I was reaching the maximum number of connections with Django (I didn't try to use the MAX_CONN_AGE, https://docs.djangoproject.com/fr/1.9/ref/settings/#conn-max-age).

**Comment by miguelgrinberg** [comment]
Great to know that you were able to make it work. What did you do for session management? That is one of the aspects that differentiates Flask-SocketIO from this project, the Flask specific package provides user session support specific to Flask on top of the base SocketIO server.

**Comment by amille44420** [comment]
For now I'm not really satisfied with but I get the session ID from the cookies we got on the client connection, from that I can get back the session. But if the session ID has to change somehow, I guess I would have to refresh that information then.

When I'll have time for it, I might find a better way to do it. For now I did develop a system similar to the Middlewares on Django. I first load my session from cookie then authenticate him after that. It's almost a real copy from the Django Middlewares.
